src/contract/event.rs

```diff
@@ -516,12 +544,103 @@ impl<T: Transport, E: ParseLog> AllEventsBuilder<T, E> {
         self
     }
 
+    /// Returns a web3 provider and filter needed for querying and streaming
+    /// events.
+    fn prepare(self) -> (Web3<T>, FilterBuilder) {
+        let mut filter_builder = self.filter.topic_filter(self.topics.compat());
+        if let Some(from_block) = self.from_block {
+            filter_builder = filter_builder.from_block(from_block);
+        }
+        if let Some(to_block) = self.to_block {
+            filter_builder = filter_builder.to_block(to_block);
+        }
+
+        (self.web3, filter_builder)
+    }
+
     /// Returns a future that resolves into a collection of events matching the
     /// event builder's parameters.
     pub fn query(self) -> QueryAllFuture<T, E> {
         QueryAllFuture::from_builder(self)
     }
 
+    /// Returns a future that resolves into a collection of events matching the
+    /// event builder's parameters. This method is similar to `query` with the
+    /// notable difference that the logs are fetched in pages by querying
+    /// smaller block ranges specified by `block_page_size` instead of using a
+    /// single query.
+    ///
+    /// # Panics
+    ///
+    /// If the block range is invalid for querying past events:
+    /// - the from block is "latest" or "pending"
+    /// - the to block is "earliest"
+    pub async fn query_past_events_paginated(self) -> Result<Vec<Event<E>>, ExecutionError> {
+        let mut start_block = match self.from_block {
+            None | Some(BlockNumber::Earliest) => 0,
+            Some(BlockNumber::Number(value)) => value.as_u64(),
+            Some(BlockNumber::Latest) | Some(BlockNumber::Pending) => {
+                panic!("invalid from block value 'latest' or 'pending'")
```

Comment:
I was just a bit lazy :stuck_out_tongue:

Reply:
The reasoning is that this is called with invalid input, not because some invariant didn't hold at runtime. I can change it.

Actually I think I will just differ to the node and let it decide what it means.

Reply:
I've decided to let the node decide what it means in this case and just forward the call to the normal query.